numba/jitclass/boxing.py

```diff
@@ -0,0 +1,198 @@
+from numba.six import exec_
+from llvmlite import ir
+import inspect
+from functools import wraps, partial
+
+_getter_code_template = """
+def accessor(__numba_self_):
+    return __numba_self_.{0}
+"""
+
+_setter_code_template = """
+def mutator(__numba_self_, __numba_val):
+    __numba_self_.{0} = __numba_val
+"""
+
+_method_code_template = """
+def method(__numba_self_, {args}):
+    return __numba_self_.{method}({args})
+"""
+
+
+def _generate_property(field, template, fname):
+    """
+    Generate simple function that get/set a field of the instance
+    """
+    source = template.format(field)
+    glbls = {}
+    exec_(source, glbls)
+    return njit(glbls[fname])
+
+
+_generate_getter = partial(_generate_property, template=_getter_code_template,
+                           fname='accessor')
+_generate_setter = partial(_generate_property, template=_setter_code_template,
+                           fname='mutator')
+
+
+def _generate_method(name, func):
+    """
+    Generate a wrapper for calling a method
+    """
+    argspec = inspect.getargspec(func)
+    assert not argspec.varargs, 'varargs not supported'
+    assert not argspec.keywords, 'keywords not supported'
+    assert not argspec.defaults, 'defaults not supported'
+
+    args = ', '.join(argspec.args[1:])  # skipped self arg
+    source = _method_code_template.format(method=name, args=args)
+    glbls = {}
+    exec_(source, glbls)
+    method = njit(glbls['method'])
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return method(*args, **kwargs)
+
+    return wrapper
+
+
+_cache_specialized_box = {}
+
+
+def _specialize_box(typ):
+    """
+    Create a subclass of Box that is specialized to the jitclass.
+
+    This function caches the result to avoid code bloat.
+    """
+    # Check cache
+    if typ in _cache_specialized_box:
+        return _cache_specialized_box[typ]
+    dct = {'__slots__': ()}
+    # Inject attributes as class properties
+    for field in typ.struct:
+        if not field.startswith('_'):
+            getter = _generate_getter(field)
+            setter = _generate_setter(field)
+            dct[field] = property(getter, setter)
+    # Inject methods as class members
+    for name, func in typ.methods.items():
+        if not name.startswith('_'):
+            getter = _generate_method(name, func)
+            dct[name] = getter
+    # Create subclass
+    subcls = type(typ.classname, (Box,), dct)
+    # Store to cache
+    _cache_specialized_box[typ] = subcls
+
+    return subcls
+
+
+class Box(object):
+    """
+    A box for numba created jit-class instance
+    """
+    __slots__ = '_meminfo', '_meminfoptr', '_dataptr', '_numba_type_'
+
+    def __new__(cls, meminfoptr, dataptr, typ):
+        if cls is not Box:
+            raise TypeError("cannot extend from specialized box")
+        return object.__new__(_specialize_box(typ))
+
+    def __init__(self, meminfoptr, dataptr, typ):
+        self._meminfo = MemInfo(meminfoptr)
```

Comment:
If `_meminfo` is only used for debug, it could be made a dynamic property (to reduce execution time of `__init__`).